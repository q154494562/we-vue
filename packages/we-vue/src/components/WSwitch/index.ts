import WSwitch from './WSwitch'

export { WSwitch }
export default WSwitch
