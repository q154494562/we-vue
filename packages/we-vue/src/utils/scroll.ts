type ElementOrWindow = HTMLElement | Window

/**
 * 找到最近的触发滚动事件的元素
 * @param {Element} element
 * @param {Element} rootParent
 * @returns {Element | window}
 */
export function getScrollEventTarget (element: HTMLElement, rootParent: ElementOrWindow = window) {
  let node = element
  while (
    node &&
    node.tagName !== 'HTML' &&
    node.tagName !== 'BODY' &&
    node.nodeType === 1 &&
    node !== rootParent
  ) {
    const { overflowY } = window.getComputedStyle(node)
    if (overflowY === 'scroll' || overflowY === 'auto') {
      return node
    }
    node = <HTMLElement>node.parentNode
  }
  return rootParent
}

// 获取滚动高度
export function getScrollTop (element: ElementOrWindow): number {
  return 'scrollTop' in element ? element.scrollTop : element.pageYOffset
}

// 设置滚动高度
export function setScrollTop (element: ElementOrWindow, value: number) {
  'scrollTop' in element ? (element.scrollTop = value) : element.scrollTo(element.scrollX, value)
}

// 获取元素距离顶部高度
export function getElementTop (element: ElementOrWindow) {
  return (
    (element === window ? 0 : (element as HTMLElement).getBoundingClientRect().top) + getScrollTop(window)
  )
}

export function getVisibleHeight (element: ElementOrWindow) {
  return element === window
    ? element.innerHeight
    : (element as HTMLElement).getBoundingClientRect().height
}
