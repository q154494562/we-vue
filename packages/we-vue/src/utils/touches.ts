function getTouch (event: TouchEvent): Touch {
  return event.changedTouches[0] || event.touches[0]
}

export { getTouch }
