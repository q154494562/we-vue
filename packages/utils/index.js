import Vue from 'vue'
import { getTouch } from './touches'

const isServer = Vue.prototype.$isServer

function isDef (value) {
  return value !== undefined && value !== null
}

function isObj (x) {
  const type = typeof x
  return x !== null && (type === 'object' || type === 'function')
}

export {
  isServer,
  getTouch,
  isDef,
  isObj
}
