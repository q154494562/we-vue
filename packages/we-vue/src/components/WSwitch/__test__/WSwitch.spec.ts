import { mount } from '@vue/test-utils'
import WSwitch from '../'
import { horizontalDrag } from '@/test/unit/utils'

const THUMB_STROKE = 20

describe('switch', () => {
  test('create', () => {
    const wrapper = mount(WSwitch, {
      propsData: {},
    })

    expect(wrapper.name()).toBe('w-switch')
    expect(wrapper.contains('.wv-switch')).toBeTruthy()

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('click', () => {
    let wrapper = mount(WSwitch, {
      propsData: {
        value: true,
      },
    })

    // first click
    wrapper.find('.wv-switch').trigger('click')
    expect(wrapper.vm.isActive).toBe(false)

    // second click
    wrapper.find('.wv-switch').trigger('click')
    expect(wrapper.vm.isActive).toBe(true)

    // click a disabled switch
    wrapper = mount(WSwitch, {
      propsData: {
        disabled: true,
        value: true,
      },
    })

    wrapper.find('.wv-switch').trigger('click')

    expect(wrapper.vm.isActive).toBe(true)
  })

  test('drag the thumb', () => {
    const wrapper = mount(WSwitch, {
      propsData: {
        value: false,
        disabled: true,
      },
    })

    // the switch can not drag when it's disabled
    horizontalDrag(wrapper.find('.thumb'), 0, THUMB_STROKE)
    expect(wrapper.vm.isActive).toBe(false)

    // enable the switch
    wrapper.setProps({
      disabled: false,
    })

    // drag to right (distance <= THUMB_STROKE / 2), the value should not change
    horizontalDrag(wrapper.find('.thumb'), 0, THUMB_STROKE / 2)
    expect(wrapper.vm.isActive).toBe(false)

    // drag to right (distance >= THUMB_STROKE / 2), the value should change
    horizontalDrag(wrapper.find('.thumb'), 0, THUMB_STROKE / 2 + 5)
    expect(wrapper.vm.isActive).toBe(true)

    // drag to left (distance <= THUMB_STROKE / 2), the value should not change
    horizontalDrag(wrapper.find('.thumb'), 0, -(THUMB_STROKE / 2))
    expect(wrapper.vm.isActive).toBe(true)

    // drag to left (distance >= THUMB_STROKE / 2), the value should change
    horizontalDrag(wrapper.find('.thumb'), 0, -(THUMB_STROKE / 2 + 5))
    expect(wrapper.vm.isActive).toBe(false)

    // drag to right with a distance bigger than THUMB_STROKE
    horizontalDrag(wrapper.find('.thumb'), 0, THUMB_STROKE + 5)
    expect(wrapper.vm.isActive).toBe(true)

    // drag to left with a distance bigger than THUMB_STROKE
    horizontalDrag(wrapper.find('.thumb'), 0, -(THUMB_STROKE + 5))
    expect(wrapper.vm.isActive).toBe(false)
  })

  test('watch isActive', () => {
    const wrapper = mount(WSwitch, {
      attachToDocument: true,
      data: function () {
        return {
          isActive: false,
        }
      },
    })

    wrapper.setData({
      isActive: true,
    })

    expect(wrapper.emitted().input).toBeTruthy()
    expect(wrapper.emitted().change).toBeTruthy()
    expect(wrapper.vm.offset).toBe(THUMB_STROKE)

    wrapper.setData({
      isActive: false,
    })
    expect(wrapper.vm.offset).toBe(0)
  })

  test('watch value', () => {
    const wrapper = mount(WSwitch, {
      propsData: {
        value: false,
      },
    })

    wrapper.setProps({
      value: false,
    })

    expect(wrapper.vm.isActive).toBe(false)

    wrapper.setProps({
      value: true,
    })

    expect(wrapper.vm.isActive).toBe(true)
  })
})
