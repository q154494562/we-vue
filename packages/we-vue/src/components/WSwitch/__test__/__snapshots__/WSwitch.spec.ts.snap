// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switch create 1`] = `
<div class="wv-switch">
  <div class="background"></div>
  <div class="thumb" style="transform: translate3d(0px, 0, 0);"></div>
</div>
`;
